dual_net.py

```diff
@@ -52,7 +59,38 @@
 
 # See www.moderndescartes.com/essays/shuffle_viz for discussion on sizing
 flags.DEFINE_integer('shuffle_buffer_size', 20000,
-                     'Size of buffer used to shuffle train examples')
+                     'Size of buffer used to shuffle train examples.')
+
+flags.DEFINE_bool('use_tpu', False, 'Whether to use TPU for training.')
+
+flags.DEFINE_string(
+    'tpu_name', None,
+    'The Cloud TPU to use for training. This should be either the name used'
+    'when creating the Cloud TPU, or a grpc://ip.address.of.tpu:8470 url.')
+
+flags.register_multi_flags_validator(
+    ['use_tpu', 'tpu_name'],
+    lambda flags: bool(flags['use_tpu']) == bool(flags['tpu_name']),
+    'If use_tpu is set, tpu_name must also be set.')
+
+flags.DEFINE_integer(
+    'iterations_per_loop', 100,
+    help=('Number of steps to run on TPU before outfeeding metrics to the CPU.'
+          ' If the number of iterations in the loop would exceed the number of'
+          ' train steps, the loop will exit before reaching'
+          ' --iterations_per_loop. The larger this value is, the higher the'
+          ' utilization on the TPU.'))
+
+flags.DEFINE_integer(
+    'num_cores', default=8,
+    help=('Number of TPU cores. For a single TPU device, this is 8 because each'
+          ' TPU has 4 chips each with 2 cores.'))
+
+flags.DEFINE_enum(
+    'precision', 'float32', ['bfloat16', 'float32'],
```

Comment:
so then what's this for?  Would we run GPU with bfloat16?  what happens if i set float32 with usetpu = true?

Reply:
gpu would probably fail unless it had bfloat16 support. TPU supports both float32 and bfloat16 mode.